ddl/ddl_worker.go

```diff
@@ -866,13 +866,27 @@ func updateSchemaVersion(t *meta.Meta, job *model.Job) (int64, error) {
 		}
 		diff.AffectedOpts = affects
 	case model.ActionTruncateTablePartition:
+		newIDs := job.Args[len(job.Args)-1].([]int64)
```

Comment:
If we add an argument in Args in a later version, len(job.Args) may get the wrong argument?

Reply:
I don't think so. This job arg is appended to `job.Args` just before `updateVerionAndTableInfo` in `partition.go`, which eventually calls `updateSchemaVersion`(e.g. [here](https://github.com/pingcap/tidb/pull/20361/files#diff-07905132777c6fce0d33bf8f27a5dfaeR1022)). As far as I know, `Args` are initialized in `ddl_api.go`, which is invoked before my new hack code. So it should always be the last argument, as long as nobody append any newer `Arg` in this dirty/hack way.

And functions that use `DecodeArgs` will work as expected due to `job.Args` are inferenced from `job.RawArgs` every invocation of `DecodeArgs`.

Reply:
It is indeed a really hack way to pass argument. It bypasses the `DecodeArgs` routines. It adds the needed argument just before calling `updateSchemaVersion`.

Reply:
I see, but this is confusing. Why not just add a field in job?

Reply:
I want to use the old style in `ApplyDiff`, which makes use of `DecodeArgs`. But it turns to be impossible. So that is where the hack comes in.

If you are to add/use a new field, should we change all code in `ApplyDiff` to that new field? Or just for the rule cache, I need a type `[2][]int` or `[]int`.

Reply:
ApplyDiff? 
`func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff)`

Reply:
Sorry, i mean `updateSchemaVersion()` and `SetSchemaDiff()`.

Reply:
I don't get your point yet. like `diff.OldTableID = job.TableID`, we can `newIDs := job.newIDs`.

Reply:
The old code use `job.DecodeArgs` to retrive extra args initialized in `ddl_api.go`(like new partition IDs for `onTruncatePartition`).

If I can retrive extra args generated in the execution process of various `onXXX` function by a new field, so can I retrive extra args initialized in `ddl_api.go`.

Or like what you said, I can add an extra field. But it is not a determined type. For truncation, I need both the old and new IDs. For drop, I need the old IDs.